File: python/leet_code/reverse_integer.py
'''
Given a signed 32-bit integer x, return x with its digits reversed. If reversing x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1], then return 0.
Assume the environment does not allow you to store 64-bit integers (signed or unsigned).

Example 1:
    Input: x = 123
    Output: 321

Example 2:
    Input: x = -123
    Output: -321

Example 3:
    Input: x = 120
    Output: 21
 
Constraints:
    -231 <= x <= 231 - 1
'''

class StringManipulationSolution:
    def reverse(self, x: int) -> int:

        max_32_bit_int = (2**31) - 1
        min_32_bit_int = -(2**31)

        
        sign = -1 if x < 0 else 1

        x = abs(x) # number is now positive

        # Convert number to string
        x_str = str(x)

        if len(x_str) <= 1:
            return x * sign
    
        reversed_digits = x_str[::-1]

        result = int(reversed_digits) * sign
        
        # print(f"Reversed digits: {reversed_digits}")
        if result > max_32_bit_int or result < min_32_bit_int:
            return 0
        
        return result

File: typescript/leet-code/reverse_integer.ts
/**
 * Given a signed 32-bit integer x, return x with its digits reversed. If reversing x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1], then return 0.
 * 
 * Assume the environment does not allow you to store 64-bit integers (signed or unsigned).
 * 
 * Example 1:
 *      Input: x = 123
 *      Output: 321
 * 
 * Example 2:
 *      Input: x = -123
 *      Output: -321
 * 
 * Example 3:
 *      Input: x = 120
 *      Output: 21
 * 
 * Constraints:
 *      -231 <= x <= 231 - 1
 */


function reverse_string_manipulation_solution(x: number): number {
  const max32Bit = 2 ** 31 - 1;
  const min32Bit = -(2 ** 31);

  const sign = x < 0 ? -1 : 1;

  x = Math.abs(x); // Ensure number is now positive

  const xStr = String(x); // Convert number to string

  if (xStr.length <= 1) {
    return x * sign;
  }

  const reversedStr = xStr.split("").reverse().join("");

  const result = Number(reversedStr) * sign;

  if (result > max32Bit || result < min32Bit) {
    return 0;
  }

  return result;
}
